File: 104-MaximumDepthofBinaryTree/104-MaximumDepthofBinaryTree.cpp
// Last updated: 9/24/2026, 8:43:46 PM
1class Solution {
2public:
3    int lastStoneWeight(vector<int>& stones) {
4        priority_queue<int> pq(stones.begin(),stones.end());
5        while(pq.size()>1){
6            int a=pq.top();
7            pq.pop();
8            int b=pq.top();
9            pq.pop();
10            if(a==b) continue;
11            pq.push(a-b);
12        }
13        if(pq.size()==0) return 0;
14        return pq.top();
15    }
16};

File: 104-MaximumDepthofBinaryTree/104-MaximumDepthofBinaryTree.java
// Last updated: 9/25/2026, 8:56:35 AM
1class Solution {
2    public void solve(int[] nums, int index, List<Integer> output, Set<List<Integer>> ans) {
3        // base case: if we have reached the end of the input array
4        if (index >= nums.length) {
5            // only add the output if it has more than one element
6            if (output.size() > 1) {
7                ans.add(new ArrayList<>(output));
8            }
9            return;
10        }
11        
12        // if the output is empty or the current element is greater than or equal to the last element in the output
13        if (output.isEmpty() || nums[index] >= output.get(output.size() - 1)) {
14            // add the current element to the output and recursively call solve
15            output.add(nums[index]);
16            solve(nums, index+1, output, ans);
17            // remove the last element from the output before returning
18            output.remove(output.size() - 1);
19        }
20        
21        // recursively call solve without adding the current element to the output
22        solve(nums, index+1, output, ans);
23    }
24    
25    public List<List<Integer>> findSubsequences(int[] nums) {
26        // use a set to store the unique sub sequences
27        Set<List<Integer>> ans = new HashSet<>();
28        solve(nums, 0, new ArrayList<>(), ans);
29        return new ArrayList<>(ans);
30    }
31}